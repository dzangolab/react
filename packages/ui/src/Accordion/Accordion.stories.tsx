import "../../public/accordion-theme.css";
import { Meta, StoryFn } from "@storybook/react";
import React from "react";

import { SubPane } from "../index";

import { Accordion } from "./index";

type AccordionType = typeof Accordion;
type AccordionProperties = Meta<AccordionType>;

const AccordionStory: AccordionProperties = {
  title: "UI/Accordion",
  component: Accordion,
};

const Template: StoryFn<AccordionType> = (arguments_) => (
  <Accordion {...arguments_}>
    <SubPane icon="https://tinyurl.com/yw5ack2m" title="Topic 1">
      Pane 1 content
    </SubPane>
    <SubPane icon="https://tinyurl.com/5cjtm2y7" title="Topic 2">
      Pane 2content
    </SubPane>
  </Accordion>
);

export const ActiveIndex: AccordionProperties = Template.bind({});
export const HorizontalDirection: AccordionProperties = Template.bind({});
export const VerticalDirection: AccordionProperties = Template.bind({});
export const CanSelfCollapse: AccordionProperties = Template.bind({});
export const CustomToggleIcons: AccordionProperties = Template.bind({});

ActiveIndex.args = {
  defaultActiveIndex: 1,
};

HorizontalDirection.args = {
  direction: "horizontal",
};

VerticalDirection.args = {
  direction: "vertical",
};

CanSelfCollapse.args = {
  canSelfCollapse: true,
};

CustomToggleIcons.args = {
  activeIcon: "https://tinyurl.com/2p8n4ee2",
  inactiveIcon: "https://tinyurl.com/6weej9xx",
};

export default AccordionStory;
