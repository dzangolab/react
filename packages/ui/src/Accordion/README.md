# Accordion component

## Installation

### npm

```
npm install @dzangolab/react-ui
```

### yarn

```
yarn add @dzangolab/react-ui
```

### pnpm

```
pnpm add @dzangolab/react-ui
```

## Usage

```
import React from "react";

import { Accordion, SubPane } from "@dzangolab/react-ui";


const Page: React.FC<Properties> = ({ children }) => {
  return (
    <Accordion
      activeIcon="chevron-down.jpg"
      inactiveIcon="chevron-up.jpg"
    >
      <SubPane icon="icon1.jpg" title="Topic 1">
        Pane 1 content
      </SubPane>
      <SubPane icon="icon2.jpg" title="Topic 2">
        Pane 2content
      </SubPane>
    </Accordion>
  );
};

export default Page
```

## Props

### Accordion

| Prop               | Type                       | Description                                      | Default    | Required |
| ------------------ | -------------------------- | ------------------------------------------------ | ---------- | -------- |
| activeIcon         | `string`                   | The toggler icon for expanded state              | None       | No       |
| canSelfCollapse    | `boolean`                  | If set to true, panes can collapse themselves    | false      | No       |
| defaultActiveIndex | `number`                   | Index of pane that should be expanded initially  | None       | No       |
| direction          | `horizontal` \| `vertical` | The direction in which the accordion is laid out | `vertical` | No       |
| inactiveIcon       | `string`                   | The toggler icon for collapsed state             | None       | No       |

### SubPane

| Prop  | Type     | Description                                     | Default | Required |
| ----- | -------- | ----------------------------------------------- | ------- | -------- |
| icon  | `string` | The icon to display in the subpane's title area | None    | No       |
| title | `string` | The text to display in the subpane's title area | None    | Yes      |
