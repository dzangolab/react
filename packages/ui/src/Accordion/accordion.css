.accordion {
  display: flex;
  list-style-type: none;
  margin: 0;
  padding: 0;
}

.accordion > li {
  display: flex;
}

.accordion > li > button {
  align-items: center;
  cursor: pointer;
  display: flex;
  flex-direction: row;
  justify-content: space-between;
  width: 100%;
}

.accordion.horizontal > li > button {
  flex-direction: column;
}

.accordion.horizontal {
  align-items: stretch;
  flex-direction: row;
}

.accordion.horizontal > li {
  flex-direction: row;
}

.accordion.horizontal > li > button {
  align-items: center;
  flex-direction: column-reverse;
  justify-content: space-between;
}

.accordion.horizontal > li > button > span {
  margin: var(--accordion-title-margin, 0 0 4px 0);
  text-orientation: sideways-right;
  transform: rotate(180deg);
  writing-mode: vertical-rl;
}

.accordion.vertical,
.accordion.vertical > li {
  flex-direction: column;
  width: 100%;
}
