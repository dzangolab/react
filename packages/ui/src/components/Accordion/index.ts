import AccordionItem from "./AccordionItem";

export { AccordionItem };
