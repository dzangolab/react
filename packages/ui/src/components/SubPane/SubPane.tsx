import React, { ReactNode } from "react";
import "./subpane.css";

type Properties = {
  children: ReactNode;
  title: string;
  icon?: string;
};

const SubPane: React.FC<Properties> = ({ children }) => {
  return <div className="sub-pane">{children}</div>;
};

export default SubPane;
