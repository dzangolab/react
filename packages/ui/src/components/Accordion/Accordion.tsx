import React, { useId, useState } from "react";

import Collapse from "../Collapse";
import AccordionTitle from "./AccordionTitle";
import "./accordion.css";

import type { ReactElement } from "react";

type Properties = {
  defaultActiveKey?: number;
  children: ReactElement[];
};

const Accordion: React.FC<Properties> = ({ defaultActiveKey, children }) => {
  const id = useId();
  const [active, setActive] = useState(defaultActiveKey);

  return (
    <ul className="accordion-list" data-testid="accordion-list">
      {children.map((item, index) => (
        <li
          className={`accordion ${active === index ? "active" : ""}`}
          key={`${id}-${index}`}
        >
          <AccordionTitle
            handleClick={setActive}
            icon={item.props.icon}
            index={index}
            title={item.props.title}
          />
          <Collapse isOpen={active === index}>{item}</Collapse>
        </li>
      ))}
    </ul>
  );
};

export default Accordion;
