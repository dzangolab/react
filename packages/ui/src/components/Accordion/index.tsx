import React, { useState } from "react";

import Collapse from "../Collapse";

import type { ReactNode } from "react";
import "./accordion.css";

interface Properties {
  accordionIcon?: ReactNode;
  children: ReactNode;
  id: string;
  label: string;
  isOpen?: boolean;
}

const Accordion = ({
  accordionIcon,
  children,
  id,
  isOpen,
  label,
}: Properties) => {
  const [open, setOpen] = useState(isOpen || false);
  const toggle = () => setOpen(!open);

  return (
    <div className="accordion">
      <button
        onClick={toggle}
        type="button"
        aria-expanded={open}
        className="accordion-trigger"
        aria-controls={`${id}-body`}
        id={id}
      >
        <span className="accordion-title">
          {label}
          {accordionIcon || <span className="accordion-icon"></span>}
        </span>
      </button>

      <div
        id={`${id}-body`}
        role="region"
        aria-labelledby={id}
        className={`accordion-panel ${!open ? "hidden" : ""}`}
      >
        <Collapse isOpen={open}>{children}</Collapse>
      </div>
    </div>
  );
};

export default Accordion;
