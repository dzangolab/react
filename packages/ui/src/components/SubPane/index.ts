import SubPane from "./SubPane";

export { SubPane };
