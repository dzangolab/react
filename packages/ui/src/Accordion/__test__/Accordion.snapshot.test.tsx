import React from "react";
import renderer, { create } from "react-test-renderer";
import { expect, test } from "vitest";

import { SubPane } from "../../components/SubPane";
import Accordion from "../Accordion";

function toJson(component: renderer.ReactTestRenderer) {
  const result = component.toJSON();

  expect(result).toBeDefined();
  expect(result).not.toBeInstanceOf(Array);

  return result as renderer.ReactTestRendererJSON;
}

test("Component matches snapshot", () => {
  const component = create(
    <Accordion activeIcon="arrow-down.jpg" inactiveIcon="arrow-up.jpg">
      <SubPane icon="icon1.jpg" title="Topic 1">
        Pane 1 content
      </SubPane>
      <SubPane icon="icon2.jpg" title="Topic 2">
        Pane 2 content
      </SubPane>
    </Accordion>,
  );
  const tree = toJson(component);
  expect(tree).toMatchSnapshot();
});
