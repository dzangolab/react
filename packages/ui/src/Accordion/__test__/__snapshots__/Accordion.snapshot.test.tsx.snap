// Vitest Snapshot v1, https://vitest.dev/guide/snapshot.html

exports[`Component matches snapshot 1`] = `
<div>
  <ul
    aria-orientation="vertical"
    class="accordion vertical"
  >
    <li
      class=""
    >
      <button
        aria-controls="pane-body-:r0:-0"
        aria-disabled="false"
        aria-expanded="false"
        aria-label="Topic 1"
        type="button"
      >
        <img
          alt="title icon"
          aria-hidden="true"
          src="icon1.jpg"
        />
        <span>
          Topic 1
        </span>
        <img
          alt="toggle icon"
          aria-hidden="true"
          src="arrow-up.jpg"
        />
      </button>
      <div
        hidden=""
        id="pane-body-:r0:-0"
        role="region"
      />
    </li>
    <li
      class=""
    >
      <button
        aria-controls="pane-body-:r0:-1"
        aria-disabled="false"
        aria-expanded="false"
        aria-label="Topic 2"
        type="button"
      >
        <img
          alt="title icon"
          aria-hidden="true"
          src="icon2.jpg"
        />
        <span>
          Topic 2
        </span>
        <img
          alt="toggle icon"
          aria-hidden="true"
          src="arrow-up.jpg"
        />
      </button>
      <div
        hidden=""
        id="pane-body-:r0:-1"
        role="region"
      />
    </li>
  </ul>
</div>
`;
